modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java

```diff
@@ -5728,6 +5746,332 @@ protected abstract Result run0(long pageId, long page, long pageAddr, BPlusIO<L>
+                                }
+                            }
+                            finally {
+                                if (nextPage != 0L)
+                                    releasePage(nextPageId, nextPage);
+                            }
+                        }
+                    }
+                    finally {
+                        if (pageAddr != 0L)
+                            readUnlock(pageId, page, pageAddr);
+                    }
+                }
+                finally {
+                    if (page != 0L)
+                        releasePage(pageId, page);
+                }
+            }
+        }
+
+        /**
+         * Removes the filtered rows from the page.
+         * Indicates the necessity to remove some row from "root".
+         *
+         * @param pageId Page id to start with.
+         * @throws IgniteCheckedException If failed.
+         */
+        private void doPurge(long pageId) throws IgniteCheckedException {
+            boolean dirty = false;
+
+            long page = acquirePage(pageId);
+            try {
+                long pageAddr = writeLock(pageId, page);
+
+                // If concurrent merge occurred we have to reinitialize cursor from the last returned row.
+                if (pageAddr == 0L) {
+                    assert lastRow != null;
+
+                    curPageId = 0L;
+
+                    return; // Retry.
+                }
+
+                try {
+                    BPlusIO<L> io = io(pageAddr);
+
+                    assert io.isLeaf();
+
+                    int cnt = io.getCount(pageAddr);
+
+                    assert cnt > 0 : cnt; // Empty leaf is nonsensical.
+
+                    dirty = modifyPage(pageId, page, null, io, pageAddr, cnt);
+                }
+                finally {
+                    writeUnlock(pageId, page, pageAddr, null, dirty);
+                }
+            }
+            finally {
+                releasePage(pageId, page);
+            }
+        }
+
+        /**
+         * Does actual modifications on the page.
+         *
+         * @param io Page IO.
+         * @param pageAddr Page address.
+         * @param cnt Item count.
+         * @return {@code true} if page was modified.
+         */
+        private boolean modifyPage(long pageId, long page, Boolean walPlc, BPlusIO<L> io, long pageAddr, int cnt)
+            throws IgniteCheckedException {
+            int write = -1;
+            int read = -1;
+            int idx = 0;
+            boolean dirty = false;
+
+            boolean h = false;
+
+            int[] idxs = null;
+            int idxsCnt = 0;
+
+            while (idx < cnt) {
+                h = clo.apply(BPlusTree.this, io, pageAddr, idx);
+
+                if (h) {
+                    if (walPlc != Boolean.TRUE && (idx != cnt - 1)) {
+                        if (idxs == null)
+                            idxs = new int[cnt - idx];
```

Comment:
Can we allocate array of `cnt` size on the start of pages iteration and reuse it?